### PR TITLE
[`WIP`] Multi-adapter saving support for PEFT

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

To be potentially merged after https://github.com/huggingface/transformers/pull/26407 
This PR adds the multi-adapter support for `save_pretrained` to be consistent with PEFT API that saves all adapters when calling `save_pretrained`. Note the default adapter is always saved in the root directory of `save_directory`.

cc @LysandreJik @BenjaminBossan @pacman100

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1995,72 +1996,67 @@ def save_pretrained(
...
+                    # the default adapter is always saved on the root directory
+                    if adapter_name != "default":
+                        current_peft_config.save_pretrained(os.path.join(save_directory, adapter_name))
+                    else:
+                        current_peft_config.save_pretrained(save_directory)
+
+                if len(peft_multi_adapter_state_dict.keys()) == 1:
+                    current_adapter = list(peft_multi_adapter_state_dict.keys())[0]
+                    state_dict = peft_multi_adapter_state_dict[current_adapter].copy()
+                    peft_multi_adapter_state_dict = None
```
Comment (BenjaminBossan) [inline src/transformers/modeling_utils.py:2028]: How about this change, which I think makes the intent a bit more obvious and avoid changing the type of `peft_multi_adapter_state_dict`:

```suggestion
                if len(peft_multi_adapter_state_dict.keys()) == 1:
                    current_adapter = list(peft_multi_adapter_state_dict.keys())[0]
                    state_dict = peft_multi_adapter_state_dict.pop(current_adapter).copy()
```

Not sure if the `.copy()` is needed?

Then, below change:

```python
- _peft_save_multi_adapter = _hf_peft_config_loaded and peft_multi_adapter_state_dict is not None
+ _peft_save_multi_adapter = _hf_peft_config_loaded and peft_multi_adapter_state_dict
```

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1995,72 +1996,67 @@ def save_pretrained(
...
-                    del state_dict[ignore_key]
+            if not _peft_save_multi_adapter:
+                for ignore_key in self._keys_to_ignore_on_save:
+                    if ignore_key in state_dict.keys():
+                        del state_dict[ignore_key]
+            else:
+                for adapter_name in peft_multi_adapter_state_dict:
+                    for ignore_key in self._keys_to_ignore_on_save:
+                        if ignore_key in peft_multi_adapter_state_dict[adapter_name].keys():
+                            del peft_multi_adapter_state_dict[adapter_name][ignore_key]
```
Comment (BenjaminBossan) [inline src/transformers/modeling_utils.py:2051]: I asked ChatGPT to make this block more elegant, here is what it came up with:

```python
if not _peft_save_multi_adapter:
    state_dict = {k: v for k, v in state_dict.items() if k not in self._keys_to_ignore_on_save}
else:
    peft_multi_adapter_state_dict = {
        adapter_name: {k: v for k, v in adapter.items() if k not in self._keys_to_ignore_on_save}
        for adapter_name, adapter in peft_multi_adapter_state_dict.items()
    }
```

WDYT? :)

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the update. In general, LGTM. I have only some minor comments, no blockers.

**Comment by BenjaminBossan** [comment]
@younesbelkada What's the status?

**Comment by younesbelkada** [comment]
@BenjaminBossan I think this might be too much of an edge case for the work it requires, I propose to keep that PR as open and if some interest arises from the community I'll work on it

**Pull request closed without merging**